### ISSUE TITLE
💡 feat: auto alert configuration with setSettings

### ISSUE THREAD

**Issue by krishtoautomate** [appium/appium#20867]
[update]
Please take a look at https://github.com/appium/appium/issues/20867#issuecomment-2614571386 comment about this issue's scope.

### Is your feature request related to a problem?

autoDismissAlerts feature is currently working with capabilities only and required only during launch and is not allowing to disable it between the tests or at certain steps as some native alert validation is important

### Describe the solution you'd like.

set settings to set  "autoDismissAlerts", true/false or "appium:autoDismissAlerts", true/false using setSettings as well:

` driver.setSetting("autoDismissAlerts", true);
   driver.setSetting("appium:autoDismissAlerts", true);`

` driver.setSetting("autoDismissAlerts", false);
   driver.setSetting("appium:autoDismissAlerts", false);`

**Comment by KazuCocoa** [comment]
When the `audoDismissAlerts` or `autoAcceptAlerts` is enabled, a new session request starts with:

```objective-c
  if (capabilities[FB_SETTING_DEFAULT_ALERT_ACTION]) {
    [FBSession initWithApplication:app
                defaultAlertAction:(id)capabilities[FB_SETTING_DEFAULT_ALERT_ACTION]];
  } else {
```
 https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Commands/FBSessionCommands.m#L213-L216

The method does:

```objective-c
+ (instancetype)initWithApplication:(nullable XCUIApplication *)application
                 defaultAlertAction:(NSString *)defaultAlertAction
{
  FBSession *session = [self.class initWithApplication:application];
  session.alertsMonitor = [[FBAlertsMonitor alloc] init];
  session.alertsMonitor.delegate = (id<FBAlertsMonitorDelegate>)session;
  session.defaultAlertAction = [defaultAlertAction lowercaseString];
  [session.alertsMonitor enable];
  return session;
}
```
https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Routing/FBSession.m#L129

When the session ends, the alert monitor will be killed like:

```objective-c
  if (nil != self.alertsMonitor) {
    [self.alertsMonitor disable];
    self.alertsMonitor = nil;
  }
```

https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Routing/FBSession.m#L141-L144

So, in this issue, we want to achieve:

1. Enable the `alertMonitor` and set `defaultAlertAction` (`accept` or `dismiss`) when the setting is configured
2. Disable the `alertMonitor`

When the settings endpoint is called, https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Commands/FBSessionCommands.m#L363 runs. It means we want to define a new settings name to configure the behavior.

**Comment by KazuCocoa** [comment]
In summary:

What we need to do here is when the settings endpoint is called, 

1. Define `autoAlerts` naming to accept autoAlerts settings in https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Utilities/FBSettings.m
2. Add a method to accept the definition in [`handleSetSettings` method](https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Commands/FBSessionCommands.m#L363)
    ```
    driver.setSetting("autoAlerts", "accept");
    driver.setSetting("autoAlerts", "dismiss");
    driver.setSetting("autoAlerts", "");  // Except for 'accept' or 'dismiss', then the auto alert will be "disabled"
    ```
3. When `autoAlerts` is acceptable words such as `accept` or `dismiss`, the 2) will enable `alertMonitor`
    - `alertMonitor` itself is currently private, so you may need to expose it in [FBSession](https://github.com/appium/WebDriverAgent/blob/c296efabe83bf9b2fcd1650c06b363344c5535e0/WebDriverAgentLib/Routing/FBSession.h) to enable/disable `alertMonitor` as instance methods.
4. Then, when `autoAlerts` is `accept` or `dismiss` -> enable `alertMonitor`. It is other words or null -> disable it

All work can be done in WebDriverAgent project.

**Comment by mykola-mokhnach** [comment]
This makes sense, although the above approach is not very flexible. In many situations accept/dismiss does not project well into alert buttons because of how WDA logic is implemented. Eventually, alerts might have 1, 2 or 3 buttons, where "accept"/"dismiss" don't make much sense to me. 

I was thinking about a possibility to provide a predicate to press a button when an alert pops up for the most flexible approach. For example:

`autoClickAlertButton: 'label =[c] allow OR label =[c] accept'` - to click either allow or accept buttons by comparing their labels case insitive

`autoClickAlertButton: 'index IN {0, 1}'` - to click either the first or the second alert button

**Comment by krishtoautomate** [comment]
Why not maintain some default names for accept/dismiss it not defined like for accept alerts : name IN {allow, yes, ok, accept} for dismiss : name IN {no, dismiss, dont}

**Comment by KazuCocoa** [comment]
I was thinking of using the existing `acceptAlertButtonSelector` / `dismissAlertButtonSelector` to set concrete rules, so this was to enable/disable auto alert behavior only.

**Comment by mykola-mokhnach** [comment]
> Why not maintain some default names for accept/dismiss if not defined, example: for **accept** alerts : name IN {allow, yes, ok, accept} for **dismiss** : name IN {no, dismiss, dont}

This is also not flexible as button names vary as well as could be localized

**Comment by mykola-mokhnach** [comment]
> I was thinking of using the existing `acceptAlertButtonSelector` / `dismissAlertButtonSelector` to set concrete rules, so this was to enable/disable auto alert behavior only to not conflict rules with them.

Above settings have different purpose. Also they use class chains rather than predicates.

**Comment by KazuCocoa** [comment]
I probably understand.

I was talking about https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Routing/FBSession.m#L54-L72 such as `acceptWithError` called from https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBAlertsMonitor.m#L58 but your suggestion was `fb_alertElement`'s detection logic in https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Utilities/FBAlertsMonitor.m#L56C34-L56C49 ?

**Comment by mykola-mokhnach** [comment]
> So, autoClickAlertPredicate will customize the alert element's detection logic itself (what kind of elements the Appium/WDA handles as 'alert') 

Not really. The alert detection logic remains the same. The given predicate must be executed inside of the alert element.


> and enable the alertsMonitor if given

yes, correct. I assume it should also have priority over alert-related capabilities

**Comment by KazuCocoa** [comment]
Gotcha, I am getting a better understanding right now. It makes sense to accept the predicate idea itself for me, btw.

Then, what about:

```
"autoClickAlert": "accept" # does the existing 'appium:autoAcceptAlerts'. It can be customized with 'acceptAlertButtonSelector'
"autoClickAlert": "dismiss" # does the existing 'appium:autoDismissAlerts'. It can be customized with 'dismissAlertButtonSelector'
"autoClickAlert": "off" # or "null". Then appium tuens it off
"autoClickAlert": "accept predict rule" # respect the predict rule over existing alert logic
```

So, `"autoClickAlert": "accept predict rule"` case, it will get a new case in https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Routing/FBSession.m#L54-L72 to handle the predicate logic. e.g. add a new method, named `acceptPredicateWithError` for instance, and it will do the given predicate over the `acceptAlertButtonSelector` / `dismissAlertButtonSelector`. To simplify, I'll create a new feature issue to add `"autoClickAlert": "accept predict rule"` case.

What about this...? Actually, my original intention in https://github.com/appium/appium/issues/20867#issuecomment-2614065854 to accept a string as a parameter was to accept a more complex rule for future implementation.

**Comment by mykola-mokhnach** [comment]
> acceptPredicateWithError

I would not use accept/dismiss in that context because it might be confusing. This predicate is supposed to click the matched predicate on an alert as soon as it is detected, whether it would be accept or dismiss or anything else is up to the user to decide based on the provided predicate.

**Comment by KazuCocoa** [comment]
Oh, year. The method naming example for WDA, when “autoClickAlert” gives predicate, was not intended. E.g. clickPredicateWithError, like that

**Comment by KazuCocoa** [comment]
Modified the naming in https://github.com/appium/appium/issues/20867#issuecomment-2614571386 as clickAlertPredicateWithError

**Comment by mykola-mokhnach** [comment]
Resolved by https://github.com/appium/appium-xcuitest-driver/pull/2550
